bundles/org.openhab.binding.icloud/src/main/java/org/openhab/binding/icloud/internal/handler/ICloudDeviceHandler.java

```diff
@@ -91,35 +94,43 @@ public void deviceInformationUpdate(List<ICloudDeviceInformation> deviceInformat
 
     @Override
     public void initialize() {
-        Bridge bridge = getBridge();
-        Object bridgeStatus = (bridge == null) ? null : bridge.getStatus();
-        logger.debug("initializeThing thing [{}]; bridge status: [{}]", getThing().getUID(), bridgeStatus);
-
         ICloudDeviceThingConfiguration configuration = getConfigAs(ICloudDeviceThingConfiguration.class);
         this.deviceId = configuration.deviceId;
 
-        ICloudAccountBridgeHandler handler = getIcloudAccount();
-        if (handler != null) {
-            refreshData();
-        } else {
-            updateStatus(OFFLINE, BRIDGE_UNINITIALIZED, "Bridge not found");
-        }
-    }
+        Bridge bridge = getBridge();
+        Object bridgeStatus = (bridge == null) ? null : bridge.getStatus();
+        this.logger.debug("initializeThing thing [{}]; bridge status: [{}]", getThing().getUID(), bridgeStatus);
 
-    private void refreshData() {
-        ICloudAccountBridgeHandler bridge = getIcloudAccount();
         if (bridge != null) {
-            bridge.refreshData();
+            ICloudAccountBridgeHandler handler = (ICloudAccountBridgeHandler) bridge.getHandler();
+            if (handler != null) {
+                handler.registerListener(this);
+                if (bridge.getStatus() == ThingStatus.ONLINE) {
+                    handler.refreshData();
+                    updateStatus(ThingStatus.ONLINE);
```

Comment:
Perhaps it would be worth considering to simply set status **UNKNOWN** here after registering as listener, and let the bridge handler override [childHandlerInitialized](https://www.openhab.org/javadoc/latest/org/openhab/core/thing/binding/bridgehandler#childHandlerInitialized(org.openhab.core.thing.binding.ThingHandler,org.openhab.core.thing.Thing)). Then bridge would call `deviceInformationUpdate` normally as soon as child handlers are ready, without this need to request the bridge for callback. It would have the benefit of only setting the things online when in fact data was received, as opposed to when data is requested (like now). WDYT?

Reply:
That would be an improvement. As i only stepped in to help out @maihacke  with the annotations/warnings, i don;t think it is up to me. If so i would need more time to understand the workings, read documentation and test and check. As the binding works and i see other bindings do similar thing i would suggest to merge it as is. During the christmass holidays i'm less time cosntraint and might comcome up with a new PR tryging to fix the remaining issues, or maybe @maihacke has other plans?

Reply:
I created a WIP issue to collect those future todos: https://github.com/openhab/openhab-addons/issues/13942